calico.go

```diff
@@ -0,0 +1,265 @@
+
+	// If running under Kubernetes then branch off into the kubernetes code, otherwise handle everything in this
+	// function.
+	if orchestratorID == "k8s" {
+		if result, err = k8s.CmdAddK8s(args, conf, hostname, etcd, endpoint); err != nil {
+			return err
+		}
+	} else {
+		// Not running in an environment that needs special casing.
+		profileID := conf.Name
+
+		if endpoint != nil {
+			// Don't create the veth or do any networking. Just update the profile on the endpoint. The profile will
+			// be created if needed during the profile processing step.
+
+			// There's an existing endpoint - no need to create another. Find the IP address from the endpoint
+			// and use that in the response.
+			fmt.Fprintf(os.Stderr, "Calico CNI appending networking profile=%s\n", profileID)
+			endpoint.ProfileIDs = append(endpoint.ProfileIDs, profileID)
+			result, err = CreateResultFromEndpoint(endpoint)
+			if err != nil {
+				return err
+			}
+		} else {
+			// There's no existing endpoint, so we need to do the following:
+			// 1) Call the configured IPAM plugin to get IP address(es)
+			// 2) Configure the calico endpoint
+			// 3) Create the veth, configuring it on both the host and container namespace.
+
+			// 1) run the IPAM plugin and make sure there's an IP address returned.
+			result, err = ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
+			if err != nil {
+				return err
+			}
+
+			// 2) Create the endpoint object
+			endpoint = &libcalico.Endpoint{ProfileIDs: []string{profileID}, OrchestratorID: orchestratorID,
+				WorkloadID: workloadID, Hostname: hostname, State: "active"}
+			if err = PopulateEndpointNets(endpoint, result); err != nil {
+				return err
+			}
+
+			fmt.Fprintf(os.Stderr, "Calico CNI using IPv4=%s IPv6=%s\n", endpoint.IPv4Nets, endpoint.IPv6Nets)
+
+			// 3) Set up the veth
+			hostVethName, contVethMac, err := DoNetworking(args, conf, result)
+			if err != nil {
+				return err
+			}
+			endpoint.Mac = contVethMac
+			endpoint.Name = hostVethName
+		}
+
+		// Write the endpoint object (either the newly created one, or the updated one with a new ProfileIDs).
+		if err := endpoint.Write(etcd); err != nil {
+			return err
+		}
+	}
+
+	// Handle profile creation - this is only done if there isn't a specific policy controller (like is sometimes
+	// used on Kubernetes)
+	if conf.Policy.PolicyType == "" {
+		// Start by checking if the profile already exists. If it already exists then there is no work to do.
+		// The CNI plugin never updates a profile.
+		exists, err := libcalico.ProfileExists(conf.Name, etcd)
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			// The profile doesn't exist so needs to be created. The rules vary depending on whether k8s is being used.
+			// Under k8s (without full policy support) the rule is permissive and allows all traffic.
+			// Otherwise, incoming traffic is only allowed from profiles with the same tag.
+			k8sInboundRule := []libcalico.Rule{libcalico.Rule{Action: "allow"}}
```

Comment:
I don't think there's a need to create `k8sInboundRule` and `tagInboundRule` in advance - just do it on lines 172 and 174.